### PR TITLE
Kinding

### PR DESCRIPTION
**Pull request by capfredf**
a kind system for typed racket

Current status: More than a "proof of concept". The new changes dovetail nicely with the existing infrastructure. Most of the existing tests pass without changes, though some tests has moved or will move from "succeed" to "fail".

Todos:
- [x] a lot of code refactoring
- [x] a RFC [rendered](https://github.com/capfredf/typed-racket/blob/kinding/rfcs/text/0003-kind-system.md)
- [x] update documentations

### CODE REVIEW

**Comment by bennn** [comment]
One of the responses to the TR survey at the last RacketCon asked for kinds:

> current typed/racket unifies polymorphic types and type constructors, so some type I can't define
> `(define-type Bar (All (A) (A Number)))`
> `(define-type Foo (Bar (All (I) (U I Symbol))))`

**Comment by capfredf** [comment]
This is a good example where higher kinds are needed, but the working branch currently doesn't support higher kinds. I will add the support down the line in a separate pull request.

File: rfcs/text/0003-kind-system.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+2. It needs to be callable such that it can be a replacement at every
+occurrence of `-pairof` where it is used to create types.
+
+A structure type with `#:property prop:procedure` specified, named
+`TypeConstructor` can easily meet those above two requirements.
+
+## Checking productivity of recursive types
+
+Previously, the type checker checked productivity when a user-provided recurive
```
Comment (capfredf) [inline rfcs/text/0003-kind-system.md:70]: clarify things

File: rfcs/text/0003-kind-system.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+Second, a new binding,`:kind`, is added for interactive uses at the Typed Racket
+ REPL.  It prints the kind of a type-level expression. For example, `(:kind
+ Integer)` will print `*` and `(:kind Listof)` will print `(=> * *)`.
+
+Lastly, programs annotated with ill-kinded types will be rejected. For example,
+when checking the program above, the type checker will report `I` in `[arg : I]`
+is not a type.
+
+# Reference-level explanation
```
Comment (capfredf) [inline rfcs/text/0003-kind-system.md:45]: separate the section into a reference-oriented part and implementation part

**Pull request merged**